[Chew Yew Keng] iP

rgonslayer opened the pull request:
# DukePro
> "Your mind is for having ideas, not holding them." - David Allen ([source](https://dansilvestre.com/productivity-quotes/))
DukePro frees your mind of having ro remember things you need to do. It's, 
- text-based
- easy to learn
- ~~FAST~~ *SUPER FAST* to use

All you need to do is, 

1. download it from [here](https://nus-cs2103-ay2223s1.github.io/website/schedule/week4/project.html).
2. double-click it. 
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:
- [x] Managing taks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

Commit:
Add Gradle support

rgonslayer committed:
Update Duke.java

Add echo for commands

rgonslayer committed:
Add Add and List Functionality

rgonslayer committed:
Add Marking

Tasks cannot be marked as done or not done yet.

Add functionality for marking and unmarking

rgonslayer committed:
Add task type

Tasks do not have a type, making it hard to tell what is the nature of the task

Add types to differentiate between different types of task

rgonslayer committed:
Add Delete feature

There is no option to delete a task from the list.

Add delete feature for users to remove a task.

rgonslayer committed:
Add error handling

Add error handling for two specific cases

rgonslayer committed:
Add Automated Text UI Testing

Testing for Jarvis is manual right now

Use a script to add automated text UI testing

rgonslayer committed:
Add Save Feature

Task List is not saved when program exits

Add auto save feature

rgonslayer committed:
Merge pull request #1 from rgonslayer/Level-7

Add Save Feature

kxrt commented on src/main/java/Jarvis.java:
```diff
@@ -0,0 +1,143 @@
+import java.util.ArrayList;
+import java.io.FileWriter;
+
+public class Jarvis {
```
In general across all the files, adding Javadoc comments would make the content clearer.

kxrt commented on src/main/java/Jarvis.java:
```diff
@@ -0,0 +1,143 @@
+        try {
+            IOhandler.readFile(taskList);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
```
Perhaps throwing a JarvisException here would be better for encapsulation.

kxrt commented on src/main/java/Jarvis.java:
```diff
@@ -0,0 +1,143 @@
+                    int toDelete = Integer.parseInt(userInput.substring(7)) - 1;
+                    Task deleteTask = taskList.get(toDelete);
+                    taskList.remove(toDelete);
+                    String deleteResponse = "Noted. I have removed this task:\n ";
+                    System.out.println(deleteResponse + deleteTask.toString());
+                    continue;
+                }
+
+                //if userinput equals to do add new to do task to list
+                if (userInput.length() > 3 && userInput.substring(0, 4).equals("todo")) {
+                    String description = userInput.substring(4);
+                    if (description.equals("")) {
+                        throw new JarvisException("The description of a todo cannot be empty");
+                    }
+                    taskList.add(new ToDo(description));
+                    System.out.println("Got it. I've added this task:\n " + taskList.get(Task.count - 1)
+                            + "\nNow you have " + (Task.count) + " tasks in the list.");
+                    continue;
+                }
+
+                //if userinput equals deadline add new deadline task to list
+                if (userInput.length() > 8 && userInput.substring(0, 8).equals("deadline")) {
+                    int divisor = userInput.indexOf("/by");
+                    String description = userInput.substring(9, divisor - 1);
+                    String date = userInput.substring(divisor + 4);
+                    taskList.add(new Deadline(description, date));
+                    System.out.println("Got it. I've added this task:\n " + taskList.get(Task.count - 1)
+                            + "\nNow you have " + (Task.count) + " tasks in the list.");
+                    continue;
+                }
+                //if userinput equals event add new event task to list
+                if (userInput.length() > 7 && userInput.substring(0, 5).equals("event")) {
+                    int divisor = userInput.indexOf("/at");
+                    String description = userInput.substring(6, divisor - 1);
+                    String date = userInput.substring(divisor + 4);
+                    taskList.add(new Event(description, date));
+                    System.out.println("Got it. I've added this task:\n " + taskList.get(Task.count - 1)
+                            + "\nNow you have " + (Task.count) + " tasks in the list.");
+                    continue;
+                }
```
Perhaps changing this to a switch statement would be better, it'd improve code readability.

kxrt commented on src/main/java/IOhandler.java:
```diff
@@ -0,0 +1,46 @@
+                String next = sc.nextLine();
+                int divisor = next.lastIndexOf("|");
+
+                if (next.charAt(0) == 'D') {
+                    String description = next.substring(8, divisor - 1);
+                    String date = next.substring((divisor + 2));
+                    taskList.add(new Deadline(description, date));
+
+                    if (next.charAt(4) == '1') {
+                        taskList.get(taskList.size() - 1).mark();
+                    }
+
+                }
+                if (next.charAt(0) == 'E') {
+                    String description = next.substring(8, divisor - 1);
+                    String date = next.substring((divisor + 2));
+                    taskList.add(new Event(description, date));
+
+                    if (next.charAt(4) == '1') {
+                        taskList.get(taskList.size() - 1).mark();
+                    }
+                }
+                if (next.charAt(0) == 'T') {
+                    String description = next.substring(8);
+                    taskList.add(new ToDo(description));
+
+                    if (next.charAt(4) == '1') {
+                        taskList.get(taskList.size() - 1).mark();
+                    }
```
I recommend abstracting hardcoded + repeated integers into a constant for better code readability.

kxrt commented on src/main/java/Jarvis.java:
```diff
@@ -0,0 +1,143 @@
+        Scanner inputScanner = new Scanner(System.in);
+
+        String userInput;
+        String introduction = "Hello. I am Jarvis \n"
+                + "What can I do for you today?";
+        String farewell = "Goodbye, have a good day.";
```
Consider changing these to constants as they remain the same across every instance of Jarvis.

kxrt commented on src/main/java/Jarvis.java:
```diff
@@ -0,0 +1,143 @@
+                FileWriter myWriter = new FileWriter(myFile);
+                for (int i = 0; i < taskList.size(); i++ ) {
+                    Task curr = taskList.get(i);
+                    if (curr instanceof Deadline) {
+                        if (curr.isDone) {
+                            myWriter.write("D" + " | 1 | " + curr.description + " | " + ((Deadline) curr).by + "\n");
+                        } else {
+                            myWriter.write("D" + " | 0 | " + curr.description + " | " + ((Deadline) curr).by + "\n");
+                        }
+                    } else if (curr instanceof Event) {
+                        if (curr.isDone) {
+                            myWriter.write("E" + " | 1 | " + curr.description + " | " + ((Event) curr).at + "\n");
+                        } else {
+                            myWriter.write("E" + " | 0 | " + curr.description + " | " + ((Event) curr).at + "\n");
+                        }
+                    } else {
+                        if (curr.isDone) {
+                            myWriter.write("T" + " | 1 | " + curr.description + "\n");
+                        } else {
+                            myWriter.write("T" + " | 0 | " + curr.description + "\n");
+                        }
+                    }
```
Perhaps changing the lowest if condition to a singular operation with a ternary operator would help avoid nesting 5 levels of code. eg. For ToDos it could be `"T | " + (curr.isDone ? "1" : "0") + " | " + curr.description + "\n"`.

kxrt commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,2 @@
+public class TaskList {
+}
```
Perhaps it would be better to shift the `List<String> taskList` from `Jarvis.java` to here for better abstraction, allowing you to define methods with cleaner code.

kxrt commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,31 @@
+public class Task {
+    protected String description;
+    protected boolean isDone;
```
Perhaps it would be better to implement a getter function for this rather than relying on a protected access modifier, as it wouldn't be effective when packages are created at the `A-Packages` stage.

kxrt commented on src/main/java/Jarvis.java:
```diff
@@ -0,0 +1,143 @@
+            }
+            System.out.print(farewell);
+        }
+        catch (JarvisException e) {
+            System.out.println((e));
```
Perhaps it would be better to abstract the running of the chatbot into a method for itself, which can instead be called here to ensure the bot continues running when facing issues.

kxrt reviewed (commented):
Quite a few recommended changes, along with the fact that Javadoc comments are completely missing. Once those are added along with the latest tags, I'm sure it'll improve a ton. Good luck!

rgonslayer committed:
Level-8

Date is stored as string

Date will be stored in a LocalDate class

rgonslayer committed:
Merge pull request #2 from rgonslayer/Level-8

Level-8

rgonslayer committed:
A-MoreOOP

Most of the code is placed in the main method

Abstract most of the code in the main method out to OOP classes.

rgonslayer committed:
A-Packages

Organize the classes into suitable java packages.

rgonslayer committed:
Merge pull request #3 from rgonslayer/A-Gradle

Add Gradle support

rgonslayer committed:
Update Task.java

rgonslayer committed:
Level-9

User can only list all items in task list currently

Add ability for User to find tasks containing a keyword

rgonslayer committed:
Merge pull request #5 from rgonslayer/Level-9

Level-9

rgonslayer committed:
A-CodingStandard

Update code according to coding standard given:

Update Try Catch to follow egyptian braces

Rest of code has followed coding standard thus far

rgonslayer committed:
A-JaveDoc

Code does not have any JavaDoc comments.

Added JavaDoc to most existing methods and classes.

rgonslayer committed:
Merge pull request #6 from rgonslayer/A-Javadoc

A-JaveDoc

rgonslayer committed:
A-CheckStyle

Add CheckStyle

rgonslayer committed:
Merge branch 'master' into A-CheckStyle

rgonslayer committed:
Merge pull request #7 from rgonslayer/A-CheckStyle

A-CheckStyle

rgonslayer committed:
Merge pull request #8 from rgonslayer/A-CodingStandard

A-CodingStandard

rgonslayer committed:
A-JUnit

Add JUnit test for existing methods and classes

rgonslayer committed:
Merge branch 'master' of https://github.com/rgonslayer/ip

rgonslayer committed:
Level-10

Add GUI for users to interact with the program instead of going through the CLI

rgonslayer committed:
Merge pull request #9 from rgonslayer/Level-10

Level-10

rgonslayer committed:
Update Introduction and Exit Condition

rgonslayer committed:
Merge pull request #10 from rgonslayer/Level-10

Update Introduction and Exit Condition

rgonslayer committed:
Create gradle.yml

There are no continuous integration processes set up yet

Thus checks and tests are done manually instead, which is repetitive and time consuming.

Continuous integration is being set up using GitHub workflows and gradle.yml

This reduces the repetitive and time consuming workflow of commits and increases productivity for developers

rgonslayer committed:
A-Assertions

There are no checks on the state of the program during execution

If there are any error in state, it might not throw any exceptions during execution

Add assertions to the program to check state as the program executes

This allows the program to verify the state as it executes so that the program executes correctly.

rgonslayer committed:
Create Ui.png

rgonslayer committed:
Improve code quality

There are long methods and arrow shaped code in the program

This reduces the readability and makes it harder for developers to maintain in the future

Refactoring code to reduce the length of methods and avoid arrow shaped code.

This will help to improve overall readability and make it easier to maintain the code in future

rgonslayer committed:
Merge pull request #12 from rgonslayer/A-CodeQuality

Improve code quality

rgonslayer committed:
Fix Checkstyle

rgonslayer committed:
Update Jarvis.java

rgonslayer committed:
Merge branch 'master' of https://github.com/rgonslayer/ip into A-Assertionss

rgonslayer committed:
Merge pull request #13 from rgonslayer/A-Assertionss

A-Assertions

rgonslayer committed:
A-UserGuide

Add UserGuide for how to use Jarvis

rgonslayer committed:
Fixed IOException error in JAR file

rgonslayer committed:
Update Storage.java

rgonslayer committed:
Refactored code

Code is failing test case as filepath was hardcoded into Storage class

Refactored code to pass test cases

rgonslayer committed:
Update gradle.yml

rgonslayer committed:
Update gradle.yml

rgonslayer committed:
Merge pull request #14 from rgonslayer/rgonslayer-patch-1

Update gradle.yml

rgonslayer committed:
Revert "Refactored code"

This reverts commit 7c152ca23b3f28c57295ea38db6045c15a418b52.

rgonslayer committed:
Revert "Revert "Refactored code""

This reverts commit 06ed8994de16bb9182eb47aee2741f1a386dd4e4.

rgonslayer committed:
Revert "Merge pull request #14 from rgonslayer/rgonslayer-patch-1"

This reverts commit 15e4abb4c3d34a761e97dfab3fc13f5f3fa0b6ef, reversing
changes made to c41239c11344be92d5fff7d8cade9f28b8c98124.

rgonslayer committed:
Revert "Update gradle.yml"

This reverts commit c41239c11344be92d5fff7d8cade9f28b8c98124.
